code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentUnitInstanceFactoryImpl.java

```diff
@@ -0,0 +1,115 @@
+    @Inject
+    ObjectProcessManager objectProcessManager;
+    @Inject
+    ServiceDiscoveryService sdService;
+    @Inject
+    LoadBalancerService lbService;
+    @Inject
+    LoadBalancerInstanceManager lbInstanceMgr;
+    @Inject
+    ObjectManager objectManager;
+    @Inject
+    ResourceMonitor resourceMonitor;
+    @Inject
+    ServiceExposeMapDao exposeMapDao;
+    @Inject
+    GenericMapDao mapDao;
+
+    @Override
+    public DeploymentUnitInstance createDeploymentUnitInstance(String uuid, Service service,
+            Pair<String, Object> objectNamePair) {
+        if (service.getKind().equalsIgnoreCase(KIND.SERVICE.name())) {
+            Pair<String, Instance> pair;
+            if (objectNamePair.getRight() != null) {
+                pair = Pair.of(objectNamePair.getLeft(), (Instance)objectNamePair.getRight());
+            } else {
+                pair = Pair.of(objectNamePair.getLeft(), null);
+            }
+            return new RegularDeploymentUnitInstance(uuid, service, pair,
+                    objectManager,
+                    resourceMonitor, objectProcessManager, sdService);
+        } else if (service.getKind().equalsIgnoreCase(KIND.LOADBALANCERSERVICE.name())) {
+            LoadBalancerHostMap hostMap = null;
+            if (objectNamePair.getRight() != null) {
+                hostMap = (LoadBalancerHostMap) objectNamePair.getRight();
+            }
+            return new LoadBalancerDeploymentUnitInstance(uuid, service,
+                    hostMap, lbService,
+                    lbInstanceMgr, objectManager, resourceMonitor, objectProcessManager, sdService);
+        }
+        return null;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<DeploymentUnitInstance> collectServiceInstances(Service service) {
+        /*
+         * find all containers related to the services through the serviceexposemaps for regular service, and
+         * loadBalancerHostMap for the lb service. Then group all the objects
+         * by the label 'io.rancher.deployment.unit'. When containers are deployed through service discovery that
+         * label will be placed on them.
+         * 
+         * If no 'io.rancher.deployment.unit' label is found on the container, generate a random UUID for the value
+         * and use that
+         */
+
+        List<DeploymentUnitInstance> deploymentUnitInstances = new ArrayList<>();
+        if (service.getKind().equalsIgnoreCase(KIND.SERVICE.name())) {
+            List<? extends Instance> serviceContainers = exposeMapDao.listServiceInstances(service.getId());
+            for (Instance serviceContainer : serviceContainers) {
+                Map<String, String> instanceLabels = DataAccessor.fields(serviceContainer)
+                        .withKey(InstanceConstants.FIELD_LABELS).as(Map.class);
+                String deploymentUnitUUID = instanceLabels.get(ServiceDiscoveryConstants.LABEL_SERVICE_DEPLOYMENT_UNIT);
```

Comment:
I always generate it per DeploymentUnit basis, not per instance basis. For deploymentUnit, its generated in DeploymentManagerImpl, addMissingUnits call